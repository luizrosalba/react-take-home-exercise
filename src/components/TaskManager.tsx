import React, { useState } from "react";

import TaskItem from "./TaskItem";
import { getInitialTasksFromLocalStorage, storeOnLocalStorage } from "../utils/localStorage";
import Button from "./Buttons/Button";

export type Task = { id: number, title: string, completed: boolean };
const TaskManager = () => {

  const [tasks, setTasks] = useState<Task[]>(getInitialTasksFromLocalStorage());
  const [filter, setFilter] = useState("all");
  const [newTask, setNewTask] = useState<string>("");

  // Intentional bug: The filter conditions are reversed.
  const filteredTasks = tasks.filter((task) => {
    if (filter === "completed") return task.completed === true;
    if (filter === "pending") return task.completed === false;
    return true;
  });

  const handleAddTask = (e: React.FormEvent) => {
    e.preventDefault();
    if (newTask) {
    if (newTask!.trim() === "") return;
      const newTaskObj = {
        id: tasks.length + 1,
        title: newTask,
        completed: false,
      };
      const updatedTasks = [...tasks, newTaskObj]
      storeOnLocalStorage({id: "storedTasks", value: updatedTasks})
      setTasks(updatedTasks);
      setNewTask("");
    }
  };

  // Intentional bug: Directly mutating the tasks array when deleting.
  const handleDeleteTask = (id: number) => {
      const task = tasks.find(task => task.id === id )
      if (task)
      {      
        const reponse = confirm(`Confirm to delete task: ${task?.title}`)
        if (reponse) {
          const updatedTasks = tasks.filter(task => task.id !== id )
          setTasks(updatedTasks);
          storeOnLocalStorage({id: "storedTasks", value: updatedTasks})
  
        }
      }
  };

  const toggleTaskCompletion = (id: number) => {
    const toggledTasks = tasks.map((task) => {
      if (task.id === id ) return {
        ...task, completed: !task.completed
      } 
      return task 
    });
    setTasks(toggledTasks)
  };

  return (
    <div className="container mx-auto bg-white p-4 rounded shadow">
      <form onSubmit={handleAddTask} className="mb-4 flex">
        <input
          type="text"
          placeholder="New task..."
          value={newTask}
          onChange={(e) => setNewTask(e.target.value)}
          className="flex-grow border rounded-l py-2 px-3"
        />
        <button type="submit" className="bg-blue-500 text-white px-4 rounded-r">
          Add
        </button>
      </form>
      <div className="flex justify-around mb-4">
        <Button 
          filter="all" 
          title={"All"} 
          setFilter={setFilter}
        /> 
        <Button 
          filter="completed" 
          title={"Completed"} 
          setFilter={setFilter}
        /> 
        <Button 
          filter="pending" 
          title={"Pending"} 
          setFilter={setFilter}
        /> 
      </div>
      <ul>
        {filteredTasks.map((task) => (
          <TaskItem
            key={task.id}
            task={task}
            onDelete={handleDeleteTask}
            onToggle={toggleTaskCompletion}
          />
        ))}
      </ul>
    </div>
  );
};

export default TaskManager;
