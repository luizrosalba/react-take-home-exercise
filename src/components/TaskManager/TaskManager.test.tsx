import '@testing-library/jest-dom'
import { fireEvent, render, screen } from '@testing-library/react'
import TaskManager from "./TaskManager"

describe('TaskManager Test', () => {
  it("Matches DOM Snapshot", () => {
    const { container } = render(<TaskManager />);
    expect(container).toMatchSnapshot();
  });
  it("Should have 2 elements as placeholders", () => {
    render(<TaskManager />);
    const element = screen.getByText('Buy groceries')
    expect(element).toBeInTheDocument();
    const element2 = screen.getByText('Clean the house')
    expect(element2).toBeInTheDocument();
  });
  it("Click on delete should delete 1 elements", () => {
    render(<TaskManager />);
    window.confirm = jest.fn().mockImplementation(() => true)
    fireEvent.click(screen.getAllByTestId('delete-button')[0])
    expect(screen.queryByText('Buy groceries')).not.toBeInTheDocument();
  });

  it("Add 1 elements", () => {
    render(<TaskManager />);
    const input = screen.getByTestId('task-input');
    fireEvent.change(input, { target: { value: 'test123' } });
    const buttonAdd = screen.getByTestId('task-add-button');
    fireEvent.click(buttonAdd)
    expect(screen.queryByText('test123')).toBeInTheDocument();
  });

});