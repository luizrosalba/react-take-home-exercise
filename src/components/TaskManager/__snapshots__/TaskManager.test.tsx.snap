// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskManager Test Matches DOM Snapshot 1`] = `
<div>
  <div
    class="container mx-auto bg-white dark:bg-gray-500 p-4 rounded shadow"
  >
    <form
      class="mb-4 flex"
    >
      <input
        class="flex-grow border rounded-l py-2 px-1 text-black bg-white dark:bg-gray-700 text-white"
        data-testid="task-input"
        placeholder="New task..."
        type="text"
        value=""
      />
      <button
        class="bg-blue-500 text-white px-4 rounded-r"
        data-testid="task-add-button"
        type="submit"
      >
        Add
      </button>
    </form>
    <div
      class="flex justify-around mb-4"
    >
      <button
        class="px-4 rounded text-gray-700 dark:text-white bg-blue-500 text-white"
      >
        All
      </button>
      <button
        class="px-4 rounded text-gray-700 dark:text-white"
      >
        Completed
      </button>
      <button
        class="px-4 rounded text-gray-700 dark:text-white"
      >
        Pending
      </button>
    </div>
    <ul>
      <li
        class="flex items-center justify-between border-b py-2"
      >
        <span
          class="cursor-pointer text-gray-900 dark:text-white"
          data-testid="toggle-button"
        >
          Buy groceries
        </span>
        <button
          class="px-2 py-2 bg-red-500 dark:bg-red-800 text-white rounded"
          data-testid="delete-button"
          style="color: white; padding: 4px 8px; border-radius: 4px;"
        >
          Delete
        </button>
      </li>
      <li
        class="flex items-center justify-between border-b py-2"
      >
        <span
          class="cursor-pointer line-through text-green-500"
          data-testid="toggle-button"
        >
          Clean the house
        </span>
        <button
          class="px-2 py-2 bg-red-500 dark:bg-red-800 text-white rounded"
          data-testid="delete-button"
          style="color: white; padding: 4px 8px; border-radius: 4px;"
        >
          Delete
        </button>
      </li>
    </ul>
  </div>
</div>
`;
